### ISSUE TITLE
Istio metrics not available on stackdriver 

### ISSUE THREAD

**Issue by nikhilbalekundargi** [GoogleCloudPlatform/istio-samples#54]
Hi,

We are installing Istio manually on GKE cluster. We want the istio metrics on Stackdriver hence we are trying to use stackdriver adapter. Followed the steps provided in this page but the istio metrics data is not available on Stackdriver. Any help would be appreciated.

**Comment by crcsmnky** [comment]
Just for completeness, can you provide the Istio version you setup and installed? And also, which steps did you follow to try and enable Istio metrics in Stackdriver?

**Comment by askmeegs** [comment]
Hi Nikhil, could you link to the install instructions you used? You have to install Istio as well as the Stackdriver adapter -- https://github.com/istio/istio/tree/master/mixer/adapter/stackdriver#mixer-stackdriver-adapter

**Comment by laurentiuspurba** [comment]
Hi Megan,

I also am having the similar issue. I don't see any metrics on GCP Trace. Looking at the link that you provided to crcmnky, I have few question:

- Do we need to set `enabled: true` on stackdriver field?
- Do we need to set the value of `apiKey:`, and `serviceAccountPath:` on auth field?
- Do we need to set `enabled: true` on tracer, and set `sampleProbability:` value to 100?

Please advise.

`mixer:
  stackdriver:
    enabled: false

  auth:
    appCredentials: false
    apiKey: ""
    serviceAccountPath: ""

  tracer:
    enabled: false
    sampleProbability: 1`

Thank you,

**Comment by nikhilbalekundargi** [comment]
Stackdriver adapter should have permission to write to Google stackdriver. Stackdriver adapter is part of Istio-telemetry. We can use Apikey to give the permission but it is deprecated now. Next option is to create a Google service account with permissions to write stackdriver metrics and logs and add the service account path to auth section and save the service account key as secrets and mount it on your pod. But Google Cloud doesn't recommend this and they have workload identity as best practice. https://cloud.google.com/kubernetes-engine/docs/how-to/workload-identity

I have resolved the issue using workload identity. Create a Google service account with metrics and tracing roles. Bind and annotate it with istio-mixer-service-account kubernets service account in istio-system namespace . The commands can be found in the link https://cloud.google.com/kubernetes-engine/docs/how-to/workload-identity.

Hope this helps

**Comment by nikhilbalekundargi** [comment]
> Hi Megan,
> 
> I also am having the similar issue. I don't see any metrics on GCP Trace. Looking at the link that you provided to crcmnky, I have few question:
> 
> * Do we need to set `enabled: true` on stackdriver field?
> * Do we need to set the value of `apiKey:`, and `serviceAccountPath:` on auth field?
> * Do we need to set `enabled: true` on tracer, and set `sampleProbability:` value to 100?
> 
> Please advise.
> 
> Thank you,
> 
> ```
>   stackdriver:
>     enabled: false
> 
>   auth:
>     appCredentials: false
>     apiKey: ""
>     serviceAccountPath: ""
> 
>   tracer:
>     enabled: false
>     sampleProbability: 1```
> ```

Stackdriver adapter should have permission to write to Google stackdriver. Stackdriver adapter is part of Istio-telemetry. We can use Apikey to give the permission but it is deprecated now. Next option is to create a Google service account with permissions to write stackdriver metrics and logs and add the service account path to auth section and save the service account key as secrets and mount it on your pod. But Google Cloud doesn't recommend this and they have workload identity as best practice. https://cloud.google.com/kubernetes-engine/docs/how-to/workload-identity

I have resolved the issue using workload identity. Create a Google service account with metrics and tracing roles. Bind and annotate it with istio-mixer-service-account kubernets service account in istio-system namespace . The commands can be found in the link https://cloud.google.com/kubernetes-engine/docs/how-to/workload-identity.

Hope this helps

**Comment by nikhilbalekundargi** [comment]
> Hi Nikhil, could you link to the install instructions you used? You have to install Istio as well as the Stackdriver adapter -- https://github.com/istio/istio/tree/master/mixer/adapter/stackdriver#mixer-stackdriver-adapter

The documentation is poor and there is no clarity on steps.

**Comment by laurentiuspurba** [comment]
> I have resolved the issue using workload identity. Create a Google service account with metrics and tracing roles. Bind and annotate it with istio-mixer-service-account kubernets service account in istio-system namespace . The commands can be found in the link https://cloud.google.com/kubernetes-engine/docs/how-to/workload-identity.
> 
> Hope this helps

Hi @nikhilbalekundargi , thanks for your response.

Before I chimed in, I have data in GCP Metrics Explorer, and I was able to create Dashboard. But there is no data coming in on GCP Trace. Other thing is when I set `enabled: true` on `tracer` field,  `stackdriver-tracing-rule` rules is created, and the rule is correct.

So, I don't know if this has something to do with workload-identity as you mentioned.

Appreciate your help.

**Comment by laurentiuspurba** [comment]
Hi @nikhilbalekundargi 
So, I set fields properties as it was before per (https://github.com/istio/istio/tree/master/mixer/adapter/stackdriver#mixer-stackdriver-adapter) as the following:

```
stackdriver:
    enabled: false

  auth:
    appCredentials: false
    apiKey: ""
    serviceAccountPath: ""

  tracer:
    enabled: false
    sampleProbability: 1
```
And now, I can see metrics on GCP Metrics explorer.

I also set the following field on my values.yaml,
`global.proxy.tracer: "stackdriver" `
 and re-ran 
`helm template ... | kubectl apply -f -`

But still, I don't see any data in Cloud Trace.

Any idea?

I appreciate your help.

**Comment by nikhilbalekundargi** [comment]
> Hi @nikhilbalekundargi
> So, I set fields properties as it was before per (https://github.com/istio/istio/tree/master/mixer/adapter/stackdriver#mixer-stackdriver-adapter) as the following:
> 
> ```
> stackdriver:
>     enabled: false
> 
>   auth:
>     appCredentials: false
>     apiKey: ""
>     serviceAccountPath: ""
> 
>   tracer:
>     enabled: false
>     sampleProbability: 1
> ```
> 
> And now, I can see metrics on GCP Metrics explorer.
> 
> I also set the following field on my values.yaml,
> `global.proxy.tracer: "stackdriver" `
> and re-ran
> `helm template ... | kubectl apply -f -`
> 
> But still, I don't see any data in Cloud Trace.
> 
> Any idea?
> 
> I appreciate your help.

@laurentiuspurba I haven't tried tracing. Sorry I cannot help you. 

As far as workload identity is concerned it is used to give permissions. I am surprised that its working without giving any credentials.

**Comment by laurentiuspurba** [comment]
> @laurentiuspurba I haven't tried tracing. Sorry I cannot help you.
> 
> As far as workload identity is concerned it is used to give permissions. I am surprised that its working without giving any credentials.

@nikhilbalekundargi I appreciate your comment on this.

**Comment by laurentiuspurba** [comment]
@crcsmnky  
cc: @nikhilbalekundargi 

Let me know if I need to open a new issue on this.

So, I was able to get Metrics Explorer (**Only Metrics Explorer**) working with stackdriver adapter with the following config:
```
stackdriver:
    enabled: false

  auth:
    appCredentials: false
    apiKey: ""
    serviceAccountPath: ""

  tracer:
    enabled: false
    sampleProbability: 1
```

In order to have Cloud Trace working, I created a Service Account as per https://github.com/istio/istio/blob/9e11075759f113c79a33801685d3bcf259b5a79f/install/kubernetes/helm/istio/charts/mixer/templates/deployment.yaml#L211.
So I tried to have Cloud Trace working, using this configuration:
```
stackdriver:
    enabled: false

  auth:
    appCredentials: false
    apiKey: ""
    serviceAccountPath: "/var/run/secrets/istio.io/telemetry/adapter/YOURSA.json"
   projectId: "YOUR-GCP-PROJECT-ID"

  tracer:
    enabled: true
    sampleProbability: 1
```

But this makes **Metrics Explorer stop receiving any data** from Istio.

Basically, I can only have either Metrics Explorer or Cloud Trace. I cannot have both.

Any idea?

**Comment by askmeegs** [comment]
Hi all - following up on this issue.
I know these issues were reported initially for Istio 1.4.x. 

I just tested Stackdriver metrics and tracing using Istio 1.6.0 and both seem to work. 
Steps to reproduce

1. created a GKE cluster.
2. installed Istio using this `IstioOperator` config. Note the [trace -> stackdriver -> `enabled:true` setting](https://github.com/askmeegs/istio-samples/blob/updates-16/common/default.yaml#L562). Also for tracing, note that the [opencensus stackdriver exporters are enabled](https://github.com/askmeegs/istio-samples/blob/updates-16/common/default.yaml#L561). And [stackdriver monitoring also set to true](https://github.com/askmeegs/istio-samples/blob/updates-16/common/default.yaml#L359).

*note that no separate Stackdriver adapter install is needed as of Istio 1.5.* 

3. Deployed a the onlineboutique sample app to the default namespace. This app has tracing instrumented with opencensus, so expect to see app-level traces in Google Cloud Trace.  
4. Opened metrics explorer - Istio generated metrics visible.

<img width="1054" alt="Screen Shot 2020-06-02 at 12 05 01 PM" src="https://user-images.githubusercontent.com/3137106/83543765-e7f28080-a4ca-11ea-9b0a-537de98de1c2.png">


5. Opened cloud trace - app-level metrics visible. 

<img width="1492" alt="Screen Shot 2020-06-02 at 12 05 33 PM" src="https://user-images.githubusercontent.com/3137106/83543772-eaed7100-a4ca-11ea-9156-6726bb2b6416.png">


I know that upgrading Istio to 1.6 right away may not be an option here, but let me know if you're continuing to see problems.

**Comment by laurentiuspurba** [comment]
@askmeegs Hi Megan, thanks for the update. This is interesting. I tried it on `Istio v1.4.3` but no luck. I might want to try again one more time.
Appreciate your input.

**Comment by askmeegs** [comment]
Hello, I'm going to close this bug since the Istio samples are now updated to 1.5 (soon 1.6) but please file an issue in the Istio repo if you keep seeing Stackdriver issues in 1.4. Thanks! 

https://github.com/istio/istio/issues
